File: bin/rules_to_bed.rb
#!/bin/env ruby

require 'optparse'
require 'ostruct'
require 'json'

### Define modules and classes here

def parse_json(filename)

    return JSON.parse(IO.readlines(filename).join)

end

### Get the script arguments and open relevant files
options = OpenStruct.new()
opts = OptionParser.new()
opts.banner = "Extracts BED formatted calling regions from json"
opts.separator ""
opts.on("-j","--json", "=JSON","JSON file") {|argument| options.json = argument }
opts.on("-h","--help","Display the usage information") {
    puts opts
    exit
}

opts.parse! 

json = parse_json(options.json)

rules = json["rules"]["bwa-freebayes"]["payload"]

rules.each do |rule|

    name = rule["name"]
    target = rule["target"].split(":")
    ref_allele = rule["matcher"].split("\t")[3]
    seq = target.shift
    from,to = target.shift.split("-").collect{|t| t.to_i }
    if to.nil?
        to = from + ref_allele.length
    end

    puts "#{seq}\t#{from-1}\t#{to+1}\t#{name}"

end

File: bin/rules_to_bed.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


import argparse
import json

parser = argparse.ArgumentParser(description="Script options")
parser.add_argument("--json", help="An rule set in JSON format")
parser.add_argument("--output")
args = parser.parse_args()


def main(json_file, output):

    # Parse the JSON file
    with open(json_file) as f:
        refs = json.load(f)

    rules = refs["rules"]["bwa-freebayes"]["payload"]

    with open(output, "w") as fo:
        for rule in rules:

            name = rule["name"]
            target = rule["target"].split(":")
            ref = rule["matcher"].split("\t")[3]
            seq = target.pop(0)
            positions = target.pop(0).split("-")
            seq_from = int(positions.pop(0))
            if not positions:
                seq_to = seq_from + len(ref)
            else:
                seq_to = int(positions.pop(0))

            fo.write(f"{seq}\t{seq_from-1}\t{seq_to+1}\t{name}\n")


if __name__ == '__main__':
    main(args.json, args.output)
